src/vos/vos_dtx.c

```diff
@@ -656,27 +670,28 @@ vos_dtx_commit_one(struct vos_container *cont, struct dtx_id *dti,
 	if (rc != 0 || epoch != 0)
 		goto out;
 
+	dtx_rec_release(cont, dae, false, &dce->dce_bcb);
+	if (dce->dce_bcb != NULL)
+		d_list_add_tail(&dce->dce_bcb_link,
+				&dce->dce_bcb->bcb_dce_list);
+
 	vos_dtx_del_cos(cont, &DAE_OID(dae), dti, DAE_DKEY_HASH(dae),
 			DAE_INTENT(dae) == DAOS_INTENT_PUNCH ? true : false);
 
-	/* XXX: Only mark the DTX as DTX_ST_COMMITTED (when commit) is not
-	 *	enough. Otherwise, some subsequent modification may change
-	 *	related data record's DTX reference or remove related data
-	 *	record as to the current DTX will have invalid reference(s)
-	 *	via its DTX record(s).
+	/* If dbtree_delete() failed, the @dae will be left in the active DTX
+	 * table until close the container. It is harmless but waste some DRAM.
 	 */
-	dtx_rec_release(umm, cont, dae, false, &dce->dce_bcb);
-	if (dce->dce_bcb != NULL)
-		d_list_add_tail(&dce->dce_bcb_link,
-				&dce->dce_bcb->bcb_dce_list);
+	dbtree_delete(cont->vc_dtx_active_hdl, BTR_PROBE_BYPASS, &kiov, NULL);
```

Comment:
I don't think DRAM btree delete can fail (unless there is some corruption).   The main thing that can cause failure with delete is the PMDK tx.

Reply:
It is not PMDK btree delete, instead, it is the modification against in-DRAM btree: we may failed when we remove DTX entry from the active DTX table (btree in DRAM) or insert DTX entry into committed DTX table (btree in DRAM).